### ISSUE TITLE
Clustering Algorithms using Quantum Computing

### ISSUE THREAD

**Issue by dquero** [Qiskit/qiskit-machine-learning#232]
With regards to ML, Qiskit 0.30 provides Quantum Neural Networks, Neural Network Classifier, Quantum Kernel Machine Learning, qGANs for Loading Random Distributions and Torch Connector and Hybrid QNNs.
I'm working to generate Clustering algorithms. At least k-means. Do you find it useful? Any suggestion?

**Comment by adekusar-drl** [comment]
Hi, certainly, this repository is a good place for quantum machine learning algorithms. If you could implement a new algorithm using Qiskit machinery, then it would be a good addition to this package. Let me know if you have questions on Qiskit!

**Comment by dquero** [comment]
I have my first version of the k-means classifier.
It does not use the euclidian distance but the cos of the angle, what is directly proportional to the distance as it can be seen in related literature.   I'm attaching the program here, I'm at your disposal for next actions.
[k-means.txt](https://github.com/Qiskit/qiskit-machine-learning/files/7455735/k-means.txt)

**Comment by adekusar-drl** [comment]
Thanks a lot. I have a few questions:
1. Is the algorithm from a paper? If so, could you please share the paper then? If no, please provide a description of the algorithm.
2. What is the advantage of using quantum computers in this algorithm?
3. Are there any comparisons versus classical algorithms?
4. The code is in the text file. If you want to finally make a contribution to QML, please take a look at the contributing guideline: https://github.com/Qiskit/qiskit-machine-learning/blob/main/CONTRIBUTING.md

**Comment by dquero** [comment]
The algorithm is one of the well known family of attempts to make clustering / non-supervised machine learning more efficient by estimating distances with quantum computing. There are many articles written about it, I even could learnt from it in Spanish. Some examples:
1. Esma Aïmeur, Gilles Brassard y Sébastien Gambs. “Aceleración cuántica para el aprendizaje no supervisado”. En: Machine Learning 90.2 (2013), pages. 261–287.
2. Nathan Wiebe, Ashish Kapoor y Krysta M. Svore. “Algoritmos cuánticos para métodos del vecino más cercano para el aprendizaje supervisado y no supervisado”. En: Cálculo de información cuántica 15.3 (2015), pages 316– 356.

In that case, the metric instead-of-distance used is P(1) of the auxiliary qbit in a cswap gate. P(1) is proportional to the euclidian distance in the Block sphere of 2 points, as demonstrated here:
https://sitiobigdata.com/2019/12/24/aprendizaje-automatico-cuantico-agrupamiento-k-means/#

My contribution to the information In this link is:
1. I provide an algorithm that estimated all the distances in parallel (that's really the advantage) instead of a loop to estimate distances.
2. Maybe it has no consequences. But the link above has a fallacy in the demonstration. It is supposed to demonstrate that the algorithm provides a value that grows as distance grows in the normal representation, but it needs to use the fact that length of the vector is one, so it really demonstrate that the value is proportional to the distance in the Bloch Sphere only.   ---> If it's interesting, I can work on demonstrating / checking that with the representation chosen the distances in the Bloch Sphere grow as they do in the normal cartesian representation if it's interesting. Intuition indicates that this is the case and it has to be true at list for the cases that distances are not very very small.

 ---

The advantage:   Instead of computing a loop with the squared root of the sum of the squares of the differences of the coordinates for each point to each centroid, it is all calculated in one step. 

----

I will work on the style.

**Comment by adekusar-drl** [comment]
Sorry for the delay. The idea, in general, is clear. There's nothing on unsupervised learning so far. The implementation may be a good contribution.
Meanwhile, could you please provide an article in English? I don't read in Spanish, unfortunately.

**Comment by prakharb10** [comment]
Hi
I recently worked on an implementation of K-Means Clustering as part of my college lab mini-project. I followed the implementation outlined here: https://arxiv.org/abs/1804.10068 pg.32-35.
I'm sharing the code with you. If it seems like a good fit, I would be willing to contribute to the module, following the guidelines
[qml.zip](https://github.com/Qiskit/qiskit-machine-learning/files/7732304/qml.zip)
.
It's a Jupyter notebook, I had to upload a zip file due to format constraints.

**Comment by adekusar-drl** [comment]
@prakharb10, Yes, a clustering algorithm can be a good contribution to Qiskit Machine Learning as we don't have anything on that implemented.

**Comment by dquero** [comment]
Hi, @prakharb10 and @adekusar-drl.
It seems that we both are following the same principle: P(1) in the control qbit of a cswap gate is a monotonous growing function with regards to the euclidiean distance of the swapped points.  As indicated above, what I do different vs what I have found is to measure all the distances to the clusters at the same time. In parallel. The fact that the numbers of qbits are expected to keep on growing and each point can be represented with one qbit only makes it possible. And this parallel calculus based on a single gate instead of all the operations involved to calculate the Euclidean distance one by one is the advantage that the quantum implementation provides. 

Here you are my code formatted with Pylint. (Only warnings in Pylint were a few variables wrongly considered constants)
I have added the extension .txt in order to be able to upload it, it can be removed and the .ipynb file is what I have executed.

[k_means_formatted.ipynb.txt](https://github.com/Qiskit/qiskit-machine-learning/files/7741068/k_means_formatted.ipynb.txt)

@prakharb10, maybe you and I can collaborate to adapt code and provide the best of the two versions. 

Thanks.
Regards.

**Comment by prakharb10** [comment]
Hi @dquero . I'm unable to open the notebook you shared. Maybe like @adekusar-drl mentioned you could share the source paper for your implementation, in English so we could peruse that.

**Comment by dquero** [comment]
Hi, I'm working to find the information in English or translate it on my own. 
With regards to the code, I always work on-line, using the IBM Quantum Lab in IBM Quantum. The file is just exported from there (formatted and added the extension .txt to be allowed to upload it here). Anyway, maybe the easiest thing is you copy it from the lines below.

"""k_means_formatted."""

# Importing standard Qiskit libraries
import numpy as np
import matplotlib.pyplot as plt
from qiskit import QuantumCircuit, execute, Aer
from qiskit.tools.jupyter import *
from qiskit.visualization import *
from ibm_quantum_widgets import *

# ---  INPUT DATA ------ (User to change them as needed)
# Points is a vector of N vectors of n elements.
# N is the dimension of the points to classify and n is the number of points.
Points = [[1, 6, 10,2,3,4,5,6,7,8,9],
          [1, 8, 10,4,3,3,7,7,7,9,9]]
# k is the number of clusters to be created
k = 2
# Centroids is a vector of N vector of k elements or an empty vector.
# (Empty vector used if user does not want to provide initial values)
#Centroids = [[1,10],[1,10]]
Centroids = []
MAX_ITERATIONS = 1000

#---  INPUT PREREQUISITES -----------
if Points == []:
    print("--- ERROR 01: Points (representing the data set) cannot be empty")
N = len(Points)
n = len(Points[0])
for i in range(N):
    if len(Points[i]) != n:
        print("--- ERROR 02: Points need to have the same number of subelements in each element")
if Centroids != []:
    if len(Centroids) != N:
        print("--- ERROR 03: Centroids must have the same number of components that Points")
    for i in range(N):
        if len(Centroids[i])!=k:
            print("--- ERROR 04: There must be k centroids.")
            print("   (Each element in Centroids is a vector of k elements).")
if N>3:
    print("--- ERROR 05: This program cannot handle more than 3 dimensions.")
    print("  Remember you can reduce dimensions losing minimal information")
    N=3
#----  PREPARATION ---------------
#Normalization of Points
Mins = np.zeros(N)
Maxs = np.zeros(N)
for i in range(N):
    Mins[i] = min(Points[i])
    Maxs[i] = max(Points[i])
NormPoints = np.zeros((N,n))
for i in range(N):
    for j in range(n):
        NormPoints[i,j] = (Points[i][j] - Mins[i]) / (Maxs[i] - Mins[i])

#Initial centroids or normalization as required
NormCentroids = np.zeros((N,k))
if Centroids == []:
    np.random.seed(100)
    for i in range(N):
        for j in range(k):
            NormCentroids[i][j]=np.random.random()
else:
    for i in range(N):
        for j in range (k):
            if (Centroids[i][j] < Mins[i] or Centroids[i][j]>Maxs[i]):
                print("--- ERROR 06: Centroids values must be in the range of the dataset")
            else:
                NormCentroids[i][j] = (Centroids[i][j] - Mins[i]) / (Maxs[i] - Mins[i])

iteration = 0
Classes = np.zeros((n))
changed = True
while ((iteration < MAX_ITERATIONS) and changed):
    iteration = iteration+1
    # ---- Calc. a correlated-to-distance-to-centroid metric for all points and centroids ---
    qc = QuantumCircuit(n*k*3,n*k)
    for i in range(n):
        for j in range(k):
            qc.rx(np.pi*NormPoints[0][i], (i*k+j)*3)
            if N>1:
                qc.rz(np.pi*NormPoints[1][i],(i*k+j)*3)
            if N>2:
                qc.ry(np.pi*NormPoints[2][i],(i*k+j)*3)

            qc.rx(np.pi*NormCentroids[0][j],(i*k+j)*3+1)
            if N>1:
                qc.rz(np.pi*NormCentroids[1][j],(i*k+j)*3+1)
            if N>2:
                qc.ry(np.pi*NormCentroids[2][j],(i*k+j)*3+1)

            qc.h((i*k+j)*3+2)
            qc.cswap((i*k+j)*3+2, (i*k+j)*3+1, (i*k+j)*3)
            qc.h((i*k+j)*3+2)

            qc.measure((i*k+j)*3+2, i*k+j)

    backend = Aer.get_backend('qasm_simulator')
    job = execute(qc, backend=backend, shots=4096)
    Keys = job.result().get_counts(qc).keys()
    Result = job.result().get_counts(qc)
    MatClasses = np.zeros((n,k))

    for key in Keys:
        for i in range(n):
            for j in range(k):
                if key[n*k-1-i*k-j]=='0':
                    MatClasses[i][j] += Result[key]

    changed = False
    for i in range(n):
        old_class = Classes[i]
        Classes[i]=0
        tmp = MatClasses[i][0]
        for j in range(k):
            if MatClasses[i][j] > tmp:
                tmp = MatClasses[i][j]
                Classes[i] = j
        if old_class != Classes[i]:
            changed = True

    NormCentroids = np.zeros((N,k))
    NumClasses = np.zeros(k)
    for i in range(n):
        tmp = int(Classes[i])
        NumClasses[tmp] = NumClasses[tmp] + 1
        for j in range(N):
            NormCentroids[j][tmp] = NormCentroids[j][tmp] + NormPoints[j][i]
    for i in range(N):
        for j in range(k):
            NormCentroids[i][j] = NormCentroids[i][j]/NumClasses[j]

# -------- Plotting -------
fig = plt.figure(figsize=(5, 5))
Colors = []
for i in range(n):
    if Classes[i]==1:
        Colors += ['red']
    else:
        Colors += ['blue']
plt.scatter(NormPoints[0], NormPoints[1], color=Colors)
plt.scatter(NormCentroids[0],NormCentroids[1])
plt.show()

**Comment by dquero** [comment]
Hi, again.
After reading carefully the article mentioned above (https://arxiv.org/abs/1804.10068) I see that the principle of the code is the same: using cswap gates to get a metric that is a monotonous growing function with regards to the euclidiean distance. And therefore, it is valid to calculate what is the nearest centroid of each point. 
What I do different is: 
   - I calculate all the point-centroid distances in parallel, while the previous algorithm does not do it, it only calculates in parallelize the dimensions.
   - I simplify to 2 dimensions if required by means of PCA. (It's a question of simplicity, keeping the outcome something easy to plot and interpretate)

I have adapted the code to the PEP8 style by using Pylint and embedded parts of the program in documented functions.
I would say it is now something reusable. See the link, as in the past, I have added the extension .txt to be allowed to upload it here. 

[k_means_formatted_v2.ipynb.txt](https://github.com/Qiskit/qiskit-machine-learning/files/7887064/k_means_formatted_v2.ipynb.txt)

Thanks.

**Comment by adekusar-drl** [comment]
@dquero Thanks a lot for the update. Let me take a look at the code and at the paper.

**Comment by dquero** [comment]
In this thread I found an empty space in quantum machine learning area of qiskit: clustering algorithms; I suggested the implementation of one of them (k-means) based on the use of cswap gates; I provided some of the papers demonstrating the algorithm is equivalent in results to the classical one; I also suggested the parallelization of the distance calculation in one shot (based on the roadmap number of qbits is expected to increase in a few years above 1M, so it will keep my version scalable).

After that, the last step is the implementation, started by @prakharb10  with a pull request.  At this point, this thread is therefore complete. Implementation, improvement, documentation, etc. are to be followed in other threads.
